### ISSUE TITLE
scalapb.validate.Validator[T] is not serialisable

### ISSUE THREAD

**Issue by Eli-Golin** [scalapb/scalapb-validate#224]
Unfortunately the [issue](https://github.com/scalapb/scalapb-validate/issues/222) is not yet fixed.

The problem is that the lack of serialisation is not related to the companion object of entity A, but rather to the Validator[A] instance  itself. The companion is never passed as a method argument (it is just part of an implicit scope), while it's public members are.

So the code:
```
object EntityValidator extends scalapb.validate.Validator[some.package.Entity] {
   def validate(input:some.package.Entity) = ....
}
```
should be serialisable.
Sorry for not making it clear enough before.

I guess you should either make it a `case object` or add and explicit `with Serializable`  to it. Which one do you prefer?

I can make a PR with the revert of a previous [change](https://github.com/scalapb/scalapb-validate/commit/f57372991c41f853b031051f48fde17be79b46cd) and introduce the new one.

**Comment by Eli-Golin** [comment]
I have validated both solutions locally,it works.

**Comment by thesamet** [comment]
Hi @Eli-Golin, prior to changing the interface (which may have future compatibility consequences), I'd like to understand better the original issue. My thinking is that since the Validator is a singleton it doesn't really need to be serialized as it is already available on all involved JVMs. Marking it serializable might enable users to unknowingly waste network resources without realizing they might have an inefficiency in their program. Will you be able to provide a minimal reproduce example, ideally as a github repo by following [these instructions](https://scalapb.github.io/docs/contact/#providing-minimal-reproducible-example)?

**Comment by Eli-Golin** [comment]
OK. I will try to provide an example (not sure about the minimalism since we are talking about a flink application after all).
It might take some time since I am in the middle of a release right now.

Regarding a singleton Validator - I think this is not the point in this case.

In Flink we have the main method which is executed on the JobManager only, and the "streaming logic" (the defined flow graph) which is executed on the TaskManagers.
On job submission to the JobManager , the "graph code" is **serialised** to the TaskManager instances for execution.

Now I have more or less the following code:

```
object main {
  def main (args:Array[String]) = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    ....
    val stream    = env.fromSource(kafkaSource, WatermarkStrategy.noWatermarks(), "1").uid("source")
    val kafkaSink = FlinkKafkaSink[Proc](topicName = outTopic, config = conf, operatorStepMsg = "Step - 5/5")
    def defineGraph(stream, kafkaSink) 
    env.execute(s"${BuildInfo.name}-$inTopic"
  }
}

object Graph { // in some common.jar
    def defineGraph[A: TypeInformation: Validator]
           (source: DataStream[(TicketInfo, KafkaHeaders)], sink: KafkaSink[(A, KafkaHeaders)]): Unit = {
                source.map(...).filter(...).flatMap (...).sinkTo(sink)
    }
}
```
So what I think happens is that the whole code inside `defineGraph` is serialised to the TaskManager instances and also the parameters to the `defineGraph` method, since it is a closure. This is the point where all parameters to the `defineGraph` method have to be serialisable.   

So just to summarise: 
The serialisation problem happens during job startup during code transfer between JobManager and TaskManagers.

P.S
As I said reproducing the problem with some minimal code is a bit complicated since we need a real flink execution environment (in this case I think a dockerised one), because the serialisation is happening between different jvm instances.
I will work on that and let you know when it is ready.

**Comment by Eli-Golin** [comment]
@thesamet Hi, 
I have eventually found the time to provide you a short example of the issue.
Please checkout [this](https://github.com/Eli-Golin/scalapb-serialization) repo.

**Comment by thesamet** [comment]
Thanks for setting up this repo, it was helpful for me to play with it. I was able to structure things in a way that Flink would be ok:

```
  def isValid[A:Validator](a:A) = Validator[A].validate(a).isSuccess

  def validate(stream:DataStream[Proc]): DataStream[Proc] = stream.filter(isValid(_))
```

I think that ultimately there should be a way to construct the graph such that the top-level function is not generic and the validator isn't part of the closure, and maybe Flink should make it a little easier to construct somehow. Anyway, this requires extra care and probably not worth the effort. I looked around and saw that some other library authors decided to make their typeclasses serializable probably for similar reasons, so I went ahead and changed it here too!
